Add breakpoint Op to Theano

carriepl opened the pull request:
This is just a small utility class to allow a user to put a conditional breakpoint, with pdb, inside the execution of a Theano function. It is implemented in a similar way to the Print() op in that it's a identity-like op with a side effect. It takes multiple inputs and the first one needs to be a 0d tensor and is interpreted as the condition that will decide the activation of the breakpoint. When, during execution, that condition is evaluated to True, the breakpoint will be activated. The subsequent inputs are interpreted as the variables the user wants to monitor and will be made available to the user when the conditional breakpoint activates.

I am not sure what the best way to test this class would be. It is easy enough to test that the breakpoint doesn't activate in a particular case but testing that it does activate seems a little bit more subtle.

abergeron commented on theano/tests/breakpoint.py:
```diff
@@ -0,0 +1,100 @@
+
+    """
+
+    __props__ = ()
```
Isn't name a property?

carriepl replied:
Good point

abergeron replied:
And by 'property' I mean something that will prevent nodes from being merged if it differs.  Which I think is the case for name here.

abergeron commented on theano/tests/breakpoint.py:
```diff
@@ -0,0 +1,100 @@
+        for i in range(len(output_storage)):
+            output_storage[i][0] = monitored[i]
+
+    def grad(self, inputs, output_gradients):
```
You might want to also add connection_pattern().

nouiz replied:
It is done.

abergeron commented:
I like this.  It can be very useful.

carriepl commented:
Multiple upgrades/bugfixes

nouiz commented on theano/tests/breakpoint.py:
```diff
@@ -0,0 +1,116 @@
+
+        # Validate that the condition is a scalar (else it is not obvious how
+        # is should be evaluated)
+        assert (condition.ndim == 0)
```
Before this line add:

if not isinstance(node, theano.Variable):
    condition = theano.tensor.as_tensor_variable(condition)

carriepl replied:
I take it you meant: " if not isinstance(condition, theano.Variable): ". Done

nouiz commented on theano/tests/breakpoint.py:
```diff
@@ -0,0 +1,116 @@
+
+        # Build the Apply node
+        inputs = [condition] + list(monitored_vars)
+        outputs = [inp.type.make_variable() for inp in monitored_vars]
```
use inp.type() instead of calling make_variable. The only difference is that inp.type() will store the stack trace when this happen. This allow Theano to print the user stack when it created this node.

carriepl replied:
Done

nouiz commented on theano/tests/test_breakpoint.py:
```diff
@@ -0,0 +1,71 @@
+
+    def test_infer_shape(self):
+
+        input1_value = numpy.arange(9).reshape(3,3).astype("float32")
```
When testing for shapes, do not use square shapes. This lower the test quality. We could mix the dimensions and the tests won't see it. Also, we have function to tests the infer_shape better then what is done it. That would be better.

carriepl replied:
Done and done. Thanks, I didn't know about InferShapeTester.

nouiz commented:
This is great, but in the current state, it won't work well with the GPU as it block the move to the GPU. Do you want to make an opt to allow the move of it to the GPU?

carriepl commented:
I changed the way the optimization works so that transfers to the GPU are propagated through the PdbBreakpoint op but it's inputs and outputs say on the CPU so that they are more easily played with by the user. In the optimization, for every pair of corresponding input and output, it ensures that if one has been moved to the GPU, so will the other. I also added an optimization to perform the same task under the new gpu backend and tests for both optimizations.

Finally, I changed the implementation of make_node following feedback from @nouiz to allow an user to reuse the same PdbBreakpoint instance on differents number of inputs. To achieve this, when make_node is called, a new instance of the op is created and used on the inputs and this instance shares the same name as the original (meaning that they will compare as equal).

carriepl commented:
Summary of IRL discussion with @lamblin : 
To minimize useless transfers between the CPU and GPU when the breakpoint is not activated but to make it easier for users to alter the content of variables on the GPU, the op should move variables to the GPU but transfer them back when the breakpoint activates so that the execution continues with the new user defined variables. This is what is implemented in this latest version.

lamblin commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1141,6 +1142,74 @@ def local_gpu_print_op(node):
+        input_transfered = []
+
+        # Propagate the transfers to gpu through the PdbBreakpoint node
+        # while leaving the PdbBreakpoint node fully on the host
```
I think that comment should be updated.

lamblin commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1141,6 +1142,74 @@ def local_gpu_print_op(node):
+            else:
+                # Both are on the gpu or on the host. No transfer is required.
+                new_inputs.append(inp)
+                input_transfered.append(False)
```
It seems to me that the case where `input_is_from_gpu and  output_goes_to_gpu` should be addressed as well.
In fact, I think it should fall in the first case, I'm not sure why `and not output_goes_to_gpu` is there.

lamblin replied:
That case may be hard to come by in a regular optimization phase, but it should be possible to make one for a test.

carriepl replied:
As for the comment, it seems I forgot to update quite a few things when I completely changed the optimizations. I'll do it right away.

lamblin commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1141,6 +1142,74 @@ def local_gpu_print_op(node):
+        # Ensure that new_op_outputs is a list of outputs (in case the op has
+        # only one output)
+        if not isinstance(new_op_outputs, list):
+            new_op_outputs = [new_op_outputs]
```
I think there is a keyword arg to `Op.__call__` so that it always returns a list, so you can skip the last couple of lines.

lamblin replied:
`new_op_outputs = node.op(*new_inputs, return_list=True)` should do the trick.

lamblin commented on theano/sandbox/gpuarray/opt.py:
```diff
@@ -331,6 +332,74 @@ def local_gpu_print_op(node):
+        input_transfered = []
+
+        # Propagate the transfers to gpu through the PdbBreakpoint node
+        # while leaving the PdbBreakpoint node fully on the host
```
All the same set of remarks for here, I guess :)

lamblin commented:
I think it's good.
@nouiz do you want to take a look before merging?

nouiz commented on theano/tests/breakpoint.py:
```diff
@@ -0,0 +1,140 @@
+
+    def perform(self, node, inputs, output_storage):
+        condition = inputs[0]
+        try:
```
I would do that only in the if condition is True. Otherwise, it will transfer all data to the CPU at each execution.

carriepl replied:
Done.

nouiz commented on theano/sandbox/gpuarray/tests/test_opt.py:
```diff
@@ -186,6 +187,25 @@ def test_print_op():
+    # breakpoint has been moved to the gpu.
+    topo = f.maker.fgraph.toposort()
+    assert isinstance(topo[-2].op, GpuElemwise)
+    assert topo[-1].op == host_from_gpu
```
This do not check that the input and output of the break point are on the GPU. If you disable that opt to move to the GPU, I think this tests still pass.

carriepl replied:
The test fails on either back-end if I disable the corresponding optimization. 

From what I understand the transfer to the gpu stems from shared variables (there are none in this graph) and the inputs of the graph. This means that if the PdbBreakpoint is not moved to the GPU, there is no way what is after could be moved to the gpu. Do I misunderstand?

nouiz replied:
right.

nouiz commented:
Seem ok for me. Anyone can merge when travis pass.
